### ISSUE TITLE
Suggestion - new C# warning

### ISSUE THREAD

**Issue by Korporal** [dotnet/roslyn#21896]
I've accidentally recently made a coding error several times, no idea why I've never noticed myself doing this before, but here it is:

Consider

```
if (some_condition)
{
   do_some_stuff();
   do_other_stuff();
}
{
   do_different_stuff();
   do_something_interesting();
}
```

as you can see an "else" is missing but it's still valid. However in this case it strikes me as better to log a warning than not. I've only found the error during testing and who knows if my testing hadn't been thorough and this was a health care application...

Of course if this were a single line **else** clause rather than a block, a warning would be a waste of time but when there is a statement or block statement after the **if** and then _another_ block statement that's **not** preceded by an **else** it is almost certainly a typo.

Maybe a more general definition is: if there is a block statement whose opening brace could be preceded by an else token and the compilation unit still be valid, then issue a warning about a possible missing else before that block statement.

**Comment by tannergooding** [comment]
I believe new warnings for existing code is generally a non-starter because it breaks `warnaserror` back-compat.

This is something that should be fairly easy to detect through the use of an analyzer, in which case you can set its diagnostic level to be info, warning, or error.

**Comment by jcouv** [comment]
I agree with Tanner. There is nothing wrong with this code from the perspective of the C# language. Analyzers are a better fit for this sort of check.

**Comment by Korporal** [comment]
@tannergooding @jcouv - Thanks, I think your suggestion makes great sense (I've never even looked at analyzers yet). Let me ask, is there already an analyzer that does this? if not can you suggest an existing example/sample that would be a good starting point for me to write one?

**Comment by jcouv** [comment]
I'm not expert in IDE projects, but I'd suggest `AddBraces` as an example.
You'll see three relevant files:
- an [analyzer](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/AddBraces/CSharpAddBracesDiagnosticAnalyzer.cs) which inspects syntax of interesting kinds and may add a diagnostic
- a [code fix](http://source.roslyn.io/#Microsoft.CodeAnalysis.CSharp.Features/AddBraces/CSharpAddBracesCodeFixProvider.cs) which triggers when such diagnostics are produced and knows how to fix the issue
- [tests](http://source.roslyn.io/#Roslyn.Services.Editor.CSharp.UnitTests/AddBraces/AddBracesTests.cs,16)

Beyond that, there is a question of whether this should be turned on by default, or be configurable. But that can be added once you have a basic design working.

**Comment by CyrusNajmabadi** [comment]
Not planning an analyzer for this.
